File: advisor/doctor.py
"""``advisor doctor`` — extended diagnostic command.

Goes beyond ``advisor status`` by also checking the surrounding environment:

* Python version (3.10+ required)
* ``git`` availability (needed for ``--since`` / ``--staged`` / ``--branch``)
* ``claude`` CLI availability (needed to actually run the pipeline)
* ``~/.claude`` directory health (exists, is a real directory, not a symlink)
* Advisor install status (nudge + skill present, current)
* Env-var detection (prints any ``ADVISOR_*`` overrides currently in effect)

Returns a :class:`DoctorReport` that can be serialized to JSON for
scripted consumption.
"""

from __future__ import annotations

import os
import platform
import shutil
import sys
from dataclasses import asdict, dataclass
from pathlib import Path
from typing import Literal

from .install import OPT_OUT_ENV, Status, get_installed_skill_version
from .install import status as install_status

# Env vars scanned by ``default_team_config`` + other advisor surfaces.
# ``OPT_OUT_ENV`` (currently ``ADVISOR_NO_NUDGE``) is appended below so this
# list stays in sync with :mod:`advisor.install` even if the opt-out env var
# is ever renamed.

HealthLevel = Literal["ok", "warn", "fail"]


@dataclass(frozen=True, slots=True)
class Check:
    """A single doctor check result."""

    name: str
    level: HealthLevel
    message: str

    def to_dict(self) -> dict[str, str]:
        return asdict(self)


@dataclass(frozen=True, slots=True)
class DoctorReport:
    """Aggregated doctor output — one check per thing inspected."""

    healthy: bool
    checks: list[Check]
    env_overrides: dict[str, str]
    advisor_version: str
    python_version: str
    platform: str

    def to_dict(self) -> dict[str, object]:
        return {
            "healthy": self.healthy,
            "checks": [c.to_dict() for c in self.checks],
            "env_overrides": self.env_overrides,
            "advisor_version": self.advisor_version,
            "python_version": self.python_version,
            "platform": self.platform,
        }


# ENV VARS scanned by default_team_config + other advisor surfaces.
# Keep ``OPT_OUT_ENV`` in the tuple so doctor auto-follows any rename.
_KNOWN_ENV_VARS = (
    "ADVISOR_MODEL",
    "ADVISOR_RUNNER_MODEL",
    "ADVISOR_MAX_RUNNERS",
    "ADVISOR_FILE_TYPES",
    "ADVISOR_MIN_PRIORITY",
    "ADVISOR_TEST_COMMAND",
    OPT_OUT_ENV,
)


def _check_python_version() -> Check:
    # pyproject.toml gates Python >= 3.10 via requires-python at install
    # time, but a broken uv environment, the wrong shim, or a manual
    # ``python3`` invocation can still bypass that and reach this code on
    # an older interpreter. Surface it as a hard fail rather than green.
    info = sys.version_info
    label = f"Python {info.major}.{info.minor}"
    # Keep this threshold in sync with `requires-python` in pyproject.toml.
    if (info.major, info.minor) < (3, 10):
        return Check("python", "fail", f"{label} — advisor requires Python 3.10+")
    return Check("python", "ok", label)


def _check_git() -> Check:
    if shutil.which("git"):
        return Check("git", "ok", "git available on PATH")
    return Check(
        "git",
        "warn",
        "git not on PATH; --since/--staged/--branch will be unavailable",
    )


def _check_claude_cli() -> Check:
    if shutil.which("claude"):
        return Check("claude-cli", "ok", "`claude` CLI available on PATH")
    return Check(
        "claude-cli",
        "warn",
        "`claude` CLI not on PATH; you cannot run the live pipeline from this shell",
    )


def _check_claude_home() -> Check:
    claude_dir = Path.home() / ".claude"
    if claude_dir.is_symlink():
        return Check(
            "claude-home",
            "warn",
            f"{claude_dir} is a symlink; advisor install refuses to write through symlinks",
        )
    if not claude_dir.exists():
        return Check(
            "claude-home",
            "warn",
            f"{claude_dir} does not exist (will be created on first `advisor install`)",
        )
    if not claude_dir.is_dir():
        return Check("claude-home", "fail", f"{claude_dir} exists but is not a directory")
    return Check("claude-home", "ok", f"{claude_dir} is a regular directory")


def _check_install(
    status: Status, installed_version: str | None, current_version: str
) -> list[Check]:
    checks: list[Check] = []
    for component in (status.nudge, status.skill):
        name = component.name
        if not component.present:
            checks.append(
                Check(
                    f"install-{name}",
                    "warn",
                    f"{name} not installed (run: advisor install)",
                )
            )
        elif not component.current:
            # Upgrade to a version-aware message when the skill badge is readable.
            if name == "skill" and installed_version and installed_version != current_version:
                checks.append(
                    Check(
                        f"install-{name}",
                        "warn",
                        f"{name} is outdated "
                        f"(installed: {installed_version}, available: {current_version}) "
                        "— run: advisor install",
                    )
                )
            else:
                checks.append(
                    Check(
                        f"install-{name}",
                        "warn",
                        f"{name} is outdated (run: advisor install)",
                    )
                )
        else:
            checks.append(Check(f"install-{name}", "ok", f"{name} installed and current"))
    if status.opt_out:
        checks.append(
            Check(
                "opt-out",
                "warn",
                f"auto-install disabled via {OPT_OUT_ENV}",
            )
        )
    return checks


def _collect_env_overrides() -> dict[str, str]:
    """Return env-var overrides currently in effect (only keys set to non-empty values)."""
    return {k: val for k in _KNOWN_ENV_VARS if (val := os.environ.get(k))}


def run_doctor(
    *,
    nudge_path: Path | None = None,
    skill_path: Path | None = None,
    version: str = "dev",
) -> DoctorReport:
    """Collect every check and return a :class:`DoctorReport`.

    Never raises — every probe is wrapped so a single broken check won't
    block the rest. The returned report is ``healthy=False`` if any check
    is ``fail``; warnings do not flip it.
    """
    quiet = os.environ.get("ADVISOR_QUIET") == "1"
    if not quiet and sys.stderr.isatty():
        from . import _style

        sys.stderr.write(_style.dim("running checks…") + "\r")
        sys.stderr.flush()
    status = install_status(nudge_path=nudge_path, skill_path=skill_path)
    installed = get_installed_skill_version(path=skill_path)
    checks: list[Check] = [
        _check_python_version(),
        _check_git(),
        _check_claude_cli(),
        _check_claude_home(),
        *_check_install(status, installed, version),
    ]
    if not quiet and sys.stderr.isatty():
        sys.stderr.write("\033[2K\r")
        sys.stderr.flush()
    healthy = not any(c.level == "fail" for c in checks)
    return DoctorReport(
        healthy=healthy,
        checks=checks,
        env_overrides=_collect_env_overrides(),
        advisor_version=version,
        python_version=platform.python_version(),
        platform=f"{platform.system()} {platform.release()}",
    )


def format_report(report: DoctorReport) -> str:
    """Human-readable (color-free) rendering of a :class:`DoctorReport`."""
    from . import _style

    lines = [
        _style.header_block(
            f"advisor doctor — {report.advisor_version}",
            [("python", report.python_version), ("platform", report.platform)],
            width=52,
        ),
        "",
    ]
    for check in report.checks:
        _label, fancy, plain, color = _style.STATE_GLYPHS[check.level]
        mark = (
            _style.paint(_style.glyph(fancy, plain), color) if color else _style.glyph(fancy, plain)
        )
        name_col = _style.paint(f"{check.name:<16}", "cyan", "bold")
        lines.append(f"  {mark} {name_col} {check.message}")
    if report.env_overrides:
        lines.append("")
        lines.append(_style.dim("  env overrides in effect:"))
        for k, v in report.env_overrides.items():
            lines.append(_style.dim(f"    {k}={v}"))
    lines.append("")
    footer = "healthy" if report.healthy else "unhealthy — fix the ✗ items above"
    color = "green" if report.healthy else "red"
    lines.append("  " + _style.paint(footer, color, "bold"))
    return "\n".join(lines)

File: src/doctor.rs
//! Port of `advisor/doctor.py` — extended diagnostic checks.

use std::path::{Path, PathBuf};

use crate::install::{
    codex_cli_available, default_codex_skills_root, get_installed_skill_version, get_status,
    update_skill_path_for, OPT_OUT_ENV,
};
use crate::version::resolve_version;

// ── known env vars ────────────────────────────────────────────────────────

const KNOWN_ENV_VARS: &[&str] = &[
    "ADVISOR_MODEL",
    "ADVISOR_RUNNER_MODEL",
    "ADVISOR_MAX_RUNNERS",
    "ADVISOR_FILE_TYPES",
    "ADVISOR_MIN_PRIORITY",
    "ADVISOR_TEST_COMMAND",
    "ADVISOR_QUIET",
    "ADVISOR_RUNNER_OUTPUT_CHAR_CEILING",
    "ADVISOR_RUNNER_FILE_READ_CEILING",
    "ADVISOR_EXPLORER_MODEL",
    "ADVISOR_MAX_EXPLORERS",
    "ADVISOR_EXPLORER_OUTPUT_CHAR_CEILING",
    "ADVISOR_EXPLORER_FILE_READ_CEILING",
    OPT_OUT_ENV,
];

/// Values for these keys are safe to surface verbatim; others are redacted.
const KNOWN_SAFE_ENV_VARS: &[&str] = &[
    "ADVISOR_MODEL",
    "ADVISOR_RUNNER_MODEL",
    "ADVISOR_MAX_RUNNERS",
    "ADVISOR_FILE_TYPES",
    "ADVISOR_MIN_PRIORITY",
    "ADVISOR_QUIET",
    "ADVISOR_RUNNER_OUTPUT_CHAR_CEILING",
    "ADVISOR_RUNNER_FILE_READ_CEILING",
    "ADVISOR_EXPLORER_MODEL",
    "ADVISOR_MAX_EXPLORERS",
    "ADVISOR_EXPLORER_OUTPUT_CHAR_CEILING",
    "ADVISOR_EXPLORER_FILE_READ_CEILING",
    OPT_OUT_ENV,
];

const REDACTED_VALUE: &str = "<set>";

// ── types ─────────────────────────────────────────────────────────────────

#[derive(Debug, Clone, PartialEq, Eq)]
pub enum HealthLevel {
    Ok,
    Warn,
    Fail,
}

impl HealthLevel {
    pub fn as_str(&self) -> &'static str {
        match self {
            Self::Ok => "ok",
            Self::Warn => "warn",
            Self::Fail => "fail",
        }
    }
}

impl std::fmt::Display for HealthLevel {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        f.write_str(self.as_str())
    }
}

#[derive(Debug, Clone)]
pub struct Check {
    pub name: String,
    pub level: HealthLevel,
    pub message: String,
}

impl Check {
    fn ok(name: &str, msg: &str) -> Self {
        Self {
            name: name.to_string(),
            level: HealthLevel::Ok,
            message: msg.to_string(),
        }
    }
    fn warn(name: &str, msg: &str) -> Self {
        Self {
            name: name.to_string(),
            level: HealthLevel::Warn,
            message: msg.to_string(),
        }
    }
    fn fail(name: &str, msg: &str) -> Self {
        Self {
            name: name.to_string(),
            level: HealthLevel::Fail,
            message: msg.to_string(),
        }
    }

    pub fn to_json(&self) -> serde_json::Value {
        serde_json::json!({
            "name": self.name,
            "level": self.level.as_str(),
            "message": self.message,
        })
    }
}

#[derive(Debug)]
pub struct DoctorReport {
    pub healthy: bool,
    pub checks: Vec<Check>,
    pub env_overrides: std::collections::HashMap<String, String>,
    pub advisor_version: String,
    pub platform: String,
}

impl DoctorReport {
    pub fn to_dict(&self) -> serde_json::Value {
        let checks: Vec<serde_json::Value> = self.checks.iter().map(|c| c.to_json()).collect();
        let env: serde_json::Map<String, serde_json::Value> = self
            .env_overrides
            .iter()
            .map(|(k, v)| (k.clone(), serde_json::json!(v)))
            .collect();
        serde_json::json!({
            "healthy": self.healthy,
            "checks": checks,
            "env_overrides": env,
            "advisor_version": self.advisor_version,
            "platform": self.platform,
        })
    }
}

// ── individual checks ─────────────────────────────────────────────────────

fn check_git() -> Check {
    if which("git") {
        Check::ok("git", "git available on PATH")
    } else {
        Check::warn(
            "git",
            "git not on PATH; --since/--staged/--branch will be unavailable",
        )
    }
}

fn check_claude_cli() -> Check {
    if which("claude") {
        Check::ok("claude-cli", "`claude` CLI available on PATH")
    } else {
        Check::warn(
            "claude-cli",
            "`claude` CLI not on PATH; Claude Code pipeline runs will be unavailable \
             from this shell — install Claude Code from https://claude.ai/code",
        )
    }
}

fn check_codex_cli() -> Check {
    if codex_cli_available() {
        Check::ok("codex-cli", "`codex` CLI available on PATH")
    } else {
        Check::warn(
            "codex-cli",
            "`codex` CLI not on PATH — Codex variant unavailable \
             (Claude Code `/advisor` is unaffected)",
        )
    }
}

fn check_home_dir(check_name: &str, dir_path: &Path) -> Check {
    if dir_path.is_symlink() {
        match std::fs::canonicalize(dir_path) {
            Ok(resolved) => {
                let home = home_dir();
                let home_resolved = std::fs::canonicalize(&home).unwrap_or(home);
                if !resolved.starts_with(&home_resolved) {
                    return Check::warn(
                        check_name,
                        &format!(
                            "{} resolves to {} (outside $HOME); \
                             advisor install refuses to write outside $HOME",
                            dir_path.display(),
                            resolved.display()
                        ),
                    );
                }
            }
            Err(e) => {
                return Check::warn(
                    check_name,
                    &format!(
                        "{} is a symlink that could not be resolved: {e}",
                        dir_path.display()
                    ),
                );
            }
        }
        if !dir_path.exists() {
            let target = std::fs::read_link(dir_path)
                .map(|p| p.display().to_string())
                .unwrap_or_else(|e| format!("<unreadable: {e}>"));
            return Check::warn(
                check_name,
                &format!(
                    "{} is a broken symlink (target: {target})",
                    dir_path.display()
                ),
            );
        }
    }
    if !dir_path.exists() {
        return Check::warn(
            check_name,
            &format!(
                "{} does not exist (will be created on first `advisor install`)",
                dir_path.display()
            ),
        );
    }
    if !dir_path.is_dir() {
        return Check::fail(
            check_name,
            &format!("{} exists but is not a directory", dir_path.display()),
        );
    }
    Check::ok(
        check_name,
        &format!("{} is a regular directory", dir_path.display()),
    )
}

fn check_claude_home() -> Check {
    check_home_dir("claude-home", &home_dir().join(".claude"))
}

fn check_codex_home() -> Check {
    check_home_dir("codex-home", &home_dir().join(".codex"))
}

fn check_codex_skill_install() -> Option<Check> {
    if !codex_cli_available() {
        return None;
    }
    let skill_path = default_codex_skills_root().join("advisor").join("SKILL.md");
    if !skill_path.exists() {
        Some(Check::warn(
            "install-codex-skill",
            &format!(
                "codex skill not installed at {} (run: advisor install)",
                skill_path.display()
            ),
        ))
    } else {
        Some(Check::ok(
            "install-codex-skill",
            &format!("codex skill installed at {}", skill_path.display()),
        ))
    }
}

fn check_harness_agent_types(status: &crate::install::Status) -> Vec<Check> {
    if status.harness_types.ok {
        vec![Check::ok(
            "harness-agent-types",
            "skill uses built-in subagent types (generalPurpose, explore)",
        )]
    } else {
        status
            .harness_types
            .issues
            .iter()
            .map(|issue| Check::warn("harness-agent-types", issue))
            .collect()
    }
}

fn check_install(
    status: &crate::install::Status,
    installed_version: Option<&str>,
    current_version: &str,
) -> Vec<Check> {
    let mut checks = Vec::new();
    let mut components: Vec<&crate::install::ComponentStatus> = vec![&status.nudge, &status.skill];
    if let Some(u) = &status.update_skill {
        components.push(u);
    }
    for component in components {
        let name = &component.name;
        if !component.present {
            checks.push(Check::warn(
                &format!("install-{name}"),
                &format!("{name} not installed (run: advisor install)"),
            ));
        } else if !component.current {
            if name == "skill" {
                if let Some(iv) = installed_version {
                    if iv != current_version {
                        checks.push(Check::warn(
                            &format!("install-{name}"),
                            &format!(
                                "{name} is outdated (installed: {iv}, available: {current_version}) \
                                 — run: advisor install"
                            ),
                        ));
                        continue;
                    }
                }
            }
            checks.push(Check::warn(
                &format!("install-{name}"),
                &format!("{name} is outdated (run: advisor install)"),
            ));
        } else {
            checks.push(Check::ok(
                &format!("install-{name}"),
                &format!("{name} installed and current"),
            ));
        }
    }
    if status.opt_out {
        checks.push(Check::warn(
            "opt-out",
            &format!("auto-install disabled via {OPT_OUT_ENV}"),
        ));
    }
    checks
}

fn collect_env_overrides() -> std::collections::HashMap<String, String> {
    let mut overrides = std::collections::HashMap::new();
    for k in KNOWN_ENV_VARS {
        if let Ok(val) = std::env::var(k) {
            let display = if KNOWN_SAFE_ENV_VARS.contains(k) {
                val
            } else {
                REDACTED_VALUE.to_string()
            };
            overrides.insert(k.to_string(), display);
        }
    }
    overrides
}

// ── public entry point ────────────────────────────────────────────────────

pub fn run_doctor(nudge_path: Option<&Path>, skill_path: Option<&Path>) -> DoctorReport {
    let version = resolve_version();
    let update_sp = update_skill_path_for(skill_path);
    let status = get_status(nudge_path, skill_path, Some(&update_sp));
    let installed = get_installed_skill_version(skill_path);

    let mut checks = vec![
        check_git(),
        check_claude_cli(),
        check_codex_cli(),
        check_claude_home(),
    ];

    let codex_present = codex_cli_available();
    if codex_present {
        checks.push(check_codex_home());
    }
    checks.extend(check_install(&status, installed.as_deref(), version));
    checks.extend(check_harness_agent_types(&status));
    if codex_present {
        if let Some(c) = check_codex_skill_install() {
            checks.push(c);
        }
    }

    let healthy = !checks.iter().any(|c| c.level == HealthLevel::Fail);
    let platform = {
        let os = std::env::consts::OS;
        os.to_string()
    };

    DoctorReport {
        healthy,
        checks,
        env_overrides: collect_env_overrides(),
        advisor_version: version.to_string(),
        platform,
    }
}

pub fn format_report(report: &DoctorReport) -> String {
    let mut lines = vec![
        format!("advisor doctor — {}", report.advisor_version),
        format!("platform: {}", report.platform),
        String::new(),
    ];
    for check in &report.checks {
        let mark = match check.level {
            HealthLevel::Ok => "✓",
            HealthLevel::Warn => "~",
            HealthLevel::Fail => "✗",
        };
        lines.push(format!("  {mark} {:<20} {}", check.name, check.message));
    }
    if !report.env_overrides.is_empty() {
        lines.push(String::new());
        lines.push("  env overrides in effect:".to_string());
        let mut pairs: Vec<_> = report.env_overrides.iter().collect();
        pairs.sort_by_key(|(k, _)| k.as_str());
        for (k, v) in pairs {
            lines.push(format!("    {k}={v:?}"));
        }
    }
    lines.push(String::new());
    let footer = if report.healthy {
        "healthy"
    } else {
        "unhealthy — fix the ✗ items above"
    };
    lines.push(format!("  {footer}"));
    lines.join("\n")
}

// ── helpers ───────────────────────────────────────────────────────────────

fn which(cmd: &str) -> bool {
    std::process::Command::new("which")
        .arg(cmd)
        .output()
        .map(|o| o.status.success())
        .unwrap_or_else(|_| {
            // Windows fallback
            std::process::Command::new("where")
                .arg(cmd)
                .output()
                .map(|o| o.status.success())
                .unwrap_or(false)
        })
}

fn home_dir() -> PathBuf {
    std::env::var("HOME")
        .or_else(|_| std::env::var("USERPROFILE"))
        .map(PathBuf::from)
        .unwrap_or_else(|_| PathBuf::from("."))
}
